### PR TITLE
admission: make admission metrics compositional

### PR DESCRIPTION
**Pull request by sttts**
Metrics emission of admission plugins and the admission chain can be implemented compositionally, i.e. completely independently from the chain logic. This PR does that, moves the whole metrics code into a sub-package to contain complexity. The plumbing logic for the emitted metrics finally is cleanly done in the apiserver bootstrapping code, instead of being totally interleaved with the core admission logic.

Ratio:
- considerably less complexity
- admission plugins are compositional, including the chain. We cannot assume that there is only one chain at the outside of the admission plugin structure. Downstream projects might have more complex admission chains, i.e. multiple chain object nested.
- addition of metrics is plumbing and should be in the apiserver plumbing code. This makes it much easier to reason about the security critical admission chain.

Follow-up of #55183 and based on #55919.

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go [hunk 1]
```diff
@@ -39,10 +39,64 @@ var (
...
+func WithMetrics(i admission.Interface, observer ObserverFunc, extraLabels ...string) admission.Interface {
+	return &pluginHandlerWithMetrics{
+		Interface:   i,
+		observer:    observer,
+		extraLabels: extraLabels,
+	}
+}
+
+// pluginHandlerWithMetrics decorates a admission handler with metrics.
+type pluginHandlerWithMetrics struct {
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go]: Nice wrap.

**Comment by sttts** [comment]
@jpbetz if you can take a look, based on your composition PR.

**Comment by deads2k** [comment]
lgtm
/approve no-issue

squash up a bit and then tag.

**Comment by deads2k** [comment]
priority/important-soon

**Comment by sttts** [comment]
> squash up a bit and then tag.

done

**Comment by jpbetz** [comment]
Thanks @sttts 

/lgtm

**Pull request merged**